### PR TITLE
Component for the newton project

### PR DESCRIPTION
**Pull request by SepidehAlassi**
solves #106

### CODE REVIEW

**Comment by SepidehAlassi** [comment]
@tobiasschweizer tests are fixed

**Comment by tobiasschweizer** [comment]
>  tests are fixed

Great! I will get the review done by the end of this week. I am excited to see content from a third party repo integrated into BEOL :-)

**Comment by SepidehAlassi** [comment]
@tobiasschweizer 😄On Friday Lukas wants to show that Newton connection works. I guess we might want to at least get the test server on this branch by tomorrow.

**Comment by subotic** [comment]
Are you going to merge into `develop` today? Deploying it on the test server shouldn't take long.

File: src/app/app.module.ts [hunk 1]
```diff
@@ -12,6 +12,8 @@ import { MaterialModule } from './material-module';
 import { FlexLayoutModule } from '@angular/flex-layout';
 // router module and app routing with all the path definitions
 // import { RouterModule } from '@angular/router';
+import {HttpClientModule} from '@angular/common/http';
```
Comment (tobiasschweizer) [inline src/app/app.module.ts]: I think this import is not used. Could this be removed?

File: src/app/resource/newton-letter/newton-letter.component.ts [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+                this.letter = text;
+            }
+        }
+    }
+    private imageSrcAttribute(element) {
+        const imgs = element.getElementsByTagName('img');
+        for (let imgIt = 0; imgIt < imgs.length; imgIt++) {
+            const image = imgs[imgIt];
+            if (image.src) {
+                image.src = image.src.replace('http://localhost:4200', 'http://www.newtonproject.ox.ac.uk');
```
Comment (tobiasschweizer) [inline src/app/resource/newton-letter/newton-letter.component.ts]: You are using `localhost:4200`. This won't work on the test server or in productive environment. Use the settings from `environment.ts` instead.

File: src/app/resource/letter/letter.component.ts [hunk 3]
```diff
@@ -76,6 +77,7 @@ export class LetterComponent extends BeolResource {
         'location': this.apiUrl + '/ontology/0801/beol/v2#location',
         'title': this.apiUrl + '/ontology/0801/beol/v2#title',
         'sysnum': this.apiUrl + '/ontology/0801/beol/v2#hasSystemNumber',
+        'npID': this.apiUrl + '/ontology/0801/newton/v2#newtonProjectID',
```
Comment (tobiasschweizer) [inline src/app/resource/letter/letter.component.ts]: Why is this necessary? Newton letters are displayed with custom components, right?

**Comment by tobiasschweizer** [comment]
@SepidehAlassi Could you please check 
https://github.com/dhlab-basel/beol/pull/108#discussion_r252578921? I think this could cause a problem on the test server.

If this is checked/fixed, please go on and deploy this branch on the test server.

File: src/app/resource/newton-letter/newton-letter.component.ts (hunk 2 above)
Comment (tobiasschweizer) [reply src/app/resource/newton-letter/newton-letter.component.ts]: Or maybe just `this.apiUrl`

**Comment by tobiasschweizer** [comment]
@SepidehAlassi Do you think we could replace cors-anywhere herokuapp with an implementation in Knora some day? 

I think this solution is perfect for the moment, but on the long run this dependency could be problematic.

**Review by tobiasschweizer — COMMENTED** [review]
I have a problem displaying the facsimiles for NATP00291: http://cudl.lib.cam.ac.uk/view/MS-ADD-09597/255 does not work (page not found)

**Comment by SepidehAlassi** [comment]
>  I think this solution is perfect for the moment, but on the long run this dependency could be problematic.
@tobiasschweizer yeah, I am aware that this can be problematic in long run, but for now it is the easy solution.

**Comment by SepidehAlassi** [comment]
> I have a problem displaying the facsimiles for NATP00291: http://cudl.lib.cam.ac.uk/view/MS-ADD-09597/255 does not work (page not found)

Currently newton xmls are not updated with the cul changes. Michael is working on it. In the next import, the URIs of the facsimiles will be updated and these errors will be solved by then.

File: src/app/resource/letter/letter.component.ts (hunk 3 above)
Comment (SepidehAlassi) [reply src/app/resource/letter/letter.component.ts]: yeah, these are cleanups which should be done. Previously the same letter component was being used for newton letters. I will clean up these idle properties soon.

File: src/app/resource/newton-letter/newton-letter.component.ts (hunk 2 above)
Comment (SepidehAlassi) [reply src/app/resource/newton-letter/newton-letter.component.ts]: done 😄

File: src/app/app.module.ts (hunk 1 above)
Comment (SepidehAlassi) [reply src/app/app.module.ts]: yes, it can be removed.

**Comment by SepidehAlassi** [comment]
@subotic we are going to merge this branch later. But the test server should be using this branch. Can you please update the test server?

**Comment by subotic** [comment]
@SepidehAlassi @tobiasschweizer done

**Comment by SepidehAlassi** [comment]
@subotic Danke 👍

**Comment by SepidehAlassi** [comment]
@subotic as a very polite, humble request: " can you please update the beol test server with this branch, please?" 😉 I have added the mathJax rendering to Newton letters, it might impress some people 😄

**Comment by subotic** [comment]
@SepidehAlassi done :-)

**Comment by SepidehAlassi** [comment]
@subotic There is a problem though with the path of the images embedded in html body.
After fetching the body of the newton lette,  I replace the environment.app URI with the newtonproject URI, locally the replacement is correct and I get the following result,

screenshot 2019-02-04 17 03 46

But the replacement of the URI fails in test server

screenshot 2019-02-04 17 05 32

is the environment.app not the one test server is using?

**Comment by kilchenmann** [comment]
environment.app is for the configuration of the Angular app url, wich can be different in each environment. When you change it into the url of the newton-server, we will have 404 errors in components or services where we need the Angular app url. The problem in the html of the Newton page is as follow: in case of an image source, it‘s not the whole url, it‘s only the relative path there... but I‘m not sure and I will have a look at it. ~~Can you send me two urls please: one from a desired source on the newton server, and one from the local beol app, where this newton source is displayed.~~ I've got it

**Comment by kilchenmann** [comment]
@SepidehAlassi I see now how you did it (yesterday it was too late):
```           
if (image.src) {
    image.src = image.src.replace(environment.app, 'http://www.newtonproject.ox.ac.uk');
    console.log(environment.app);
}
```

Nice idea but I'm not sure if that's the best practice. But we will figure out, how we can solve it 😉

**Comment by kilchenmann** [comment]
@subotic are you using the config.dev.json for the deployment? We should try to fix the app configuration issue here in beol... What do you think?

**Comment by subotic** [comment]
no, I have two custom environments which I copy into the source code. Afterwards I build the app. Maybe we could change BEOL to use config.xxx.json? They don’t use the firestore stuff anymore. Then @SepidehAlassi could easily change the config at runtime.

**Comment by SepidehAlassi** [comment]
@kilchenmann and @subotic Thank you very much guys for thinking about this problem, I'll appreciate any solution (dirty or neat 😉 )

**Comment by kilchenmann** [comment]
@subotic can we do that together? next Monday and in a separate branch?

**Comment by subotic** [comment]
yes, sure.

**Comment by tobiasschweizer** [comment]
@SepidehAlassi I propose the following procedure:

- I will review this PR
- once merged, we deploy `develop` on the test server
- deploy on live server

Is that ok? I will do the review as soon as this PR is ready.

**Comment by SepidehAlassi** [comment]
@tobiasschweizer The problem which Andre and Ivan are working on should be solved first before we merge this branch to develop.

**Comment by tobiasschweizer** [comment]
> The problem which Andre and Ivan are working on should be solved first before we merge this branch to develop.

I understand. How much time do they need appr.?

**Comment by SepidehAlassi** [comment]
> I understand. How much time do they need appr.?

not sure, they had planned to work on it this week.
@subotic & @kilchenmann any update/progress with config setup for beol?

**Comment by subotic** [comment]
@SepidehAlassi we got it "almost" working yesterday. I hope @kilchenmann can find some time this week to finish it.

**Comment by SepidehAlassi** [comment]
> @SepidehAlassi we got it "almost" working yesterday. I hope @kilchenmann can find some time this week to finish it.

@subotic & @kilchenmann great, thank you guys. 😄

**Comment by kilchenmann** [comment]
> @subotic & @kilchenmann any update/progress with config setup for beol?

It's done and works in knora-ui modules. @subotic is testing it in his own app and then we can use the same setup in the beol app.

**Comment by SepidehAlassi** [comment]
> It's done and works in knora-ui modules. @subotic is testing it in his own app and then we can use the same setup in the beol app.

That is great news. Thank you very much.

**Comment by tobiasschweizer** [comment]
There is still a problem for the productive build: https://github.com/dhlab-basel/Knora-ui/issues/178

**Comment by SepidehAlassi** [comment]
@tobiasschweizer this can be reviewed now, you can use the old data for it.

**Comment by tobiasschweizer** [comment]
@SepidehAlassi ok, I will do it tomorrow.

**Comment by SepidehAlassi** [comment]
> ok, I will do it tomorrow.

@tobiasschweizer thanks

**Comment by SepidehAlassi** [comment]
@tobiasschweizer sorry, I just remembered you can not use the old data to check this. I have uploaded new data in switchdrive, please use the data there to check this pull request.

File: src/app/resource/newton-letter/newton-letter.component.ts [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+    // this is for our own (knora) resources
+    initProps() {
+
+        const props = new LetterProps();
+
+        this.mapper(props);
+        this.props = props;
+        // get the id from the route newtonletter/:id e.g. NATP00120
+        this.navigationSubscription = this._route.paramMap.subscribe((params: ParamMap) => {
```
Comment (tobiasschweizer) [inline src/app/resource/newton-letter/newton-letter.component.ts]: I think it is not necessary to subscribe to `paramMap` here since this is handled in `BEOLResource`. `initProps` is called when everything is ready.

Comment (tobiasschweizer) [reply src/app/resource/newton-letter/newton-letter.component.ts]: It should be enough to just do `this.getNewtonLetterText(this.props.npID[0].getContent());`

File: src/assets/style/_elements.scss [hunk 5]
```diff
@@ -20,8 +20,8 @@
 
 .main-content {
   margin: 0 auto 0 $sidenav-width;
-  max-width: 900px;
-  padding: 0 48px 24px 48px;
+  max-width: 1000px;
```
Comment (tobiasschweizer) [inline src/assets/style/_elements.scss:23]: I am not sure if that conflicts with some styling added by @kilchenmann for the search bar.

File: src/app/pipes/sanitize-html.pipe.ts [hunk 6]
```diff
@@ -0,0 +1,15 @@
+import { Pipe, PipeTransform } from '@angular/core';
+import { DomSanitizer, SafeHtml } from '@angular/platform-browser';
+
+@Pipe({
+    name: 'sanitizeHtml'
+})
+export class SanitizeHtmlPipe implements PipeTransform {
```
Comment (tobiasschweizer) [inline src/app/pipes/sanitize-html.pipe.ts:7]: Where does this get used in `newton-letter.component.ts`?

File: src/app/pipes/sanitize-url.pipe.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import { Pipe, PipeTransform } from '@angular/core';
+import { DomSanitizer, SafeResourceUrl, SafeUrl } from '@angular/platform-browser';
+
+@Pipe({
+    name: 'sanitizeUrl'
+})
+export class SanitizeUrlPipe implements PipeTransform {
```
Comment (tobiasschweizer) [inline src/app/pipes/sanitize-url.pipe.ts:7]: Where does this get used in `newton-letter.component.ts`?

File: src/app/resource/newton-letter/newton-letter.component.html [hunk 8]
```diff
@@ -0,0 +1,129 @@
+<div class="full-size scroll-y-only">
+
+    <kui-progress-indicator *ngIf="isLoading" [color]="'#D88958'"></kui-progress-indicator>
```
Comment (tobiasschweizer) [inline src/app/resource/newton-letter/newton-letter.component.html:3]: I don't see the progress indicator while the letter is loading. Maybe it is the parsing the takes up the time?

Comment (SepidehAlassi) [reply src/app/resource/newton-letter/newton-letter.component.html:3]: probably because it is the communication with the newton server which takes time.

File: src/assets/style/_elements.scss (hunk 5 above)
Comment (SepidehAlassi) [reply src/assets/style/_elements.scss:23]: If there is a conflict we can then adjust this.

File: src/app/resource/newton-letter/newton-letter.component.html (hunk 8 above)
Comment (tobiasschweizer) [reply src/app/resource/newton-letter/newton-letter.component.html:3]: but you set the `isLoading` variable correctly as it seems to me

File: src/assets/style/_elements.scss (hunk 5 above)
Comment (tobiasschweizer) [reply src/assets/style/_elements.scss:23]: @kilchenmann would know ;-)

File: src/app/pipes/sanitize-html.pipe.ts (hunk 6 above)
Comment (SepidehAlassi) [reply src/app/pipes/sanitize-html.pipe.ts:7]: sanitizeHtml was used in newton-letter.component.html, as

`<div *ngIf="!isLoading && letter"  class="main-content without-sidebar">
        <div [innerHTML]="letter | sanitizeHtml"></div>
    </div>`

 But I removed it when I fed the letter to mathJax for rendering, and changed the above part to
`  <div *ngIf="!isLoading && letter"  class="main-content without-sidebar">
                <div [mathJax]="letter" [bindEvents]="false" [renderMath]="true"></div>
            </div>`

Since then there has been no error with the html files retrieved from Newton. I still have not got the black magic behind mathJax, but using mathJax to display letter we do not need the sanitizer anymore

File: src/app/resource/newton-letter/newton-letter.component.html (hunk 8 above)
Comment (SepidehAlassi) [reply src/app/resource/newton-letter/newton-letter.component.html:3]: isLoading is used for just loading the metadata of the letters which are imported in knora. Since the text is fetched from newton server on the fly and is not a response of knora, you do not see the progress indicator for loading the text of the newton letters

File: src/app/pipes/sanitize-url.pipe.ts (hunk 7 above)
Comment (SepidehAlassi) [reply src/app/pipes/sanitize-url.pipe.ts:7]: see the reply above.

File: src/app/resource/newton-letter/newton-letter.component.ts (hunk 4 above)
Comment (SepidehAlassi) [reply src/app/resource/newton-letter/newton-letter.component.ts]: well `this.navigationSubscription` is of class type subscription, but 
`this.getNewtonLetterText(this.props.npID[0].getContent());`
returns void. I can not apply your suggestion because of the difference of types.
Therefore for mapping reason, I am basically using the subscription to paramMap. I am not an expert in type handling of typescript, hence, if you know another way of this type-handling please do let me know.

File: src/app/pipes/sanitize-html.pipe.ts (hunk 6 above)
Comment (SepidehAlassi) [reply src/app/pipes/sanitize-html.pipe.ts:7]: I am not sure though if we need to get rid of the sanitizer, as I am with high probability going to need it for the text search results communicated with the newton project

File: src/app/resource/newton-letter/newton-letter.component.ts (hunk 4 above)
Comment (tobiasschweizer) [reply src/app/resource/newton-letter/newton-letter.component.ts]: I don't understand.  Couldn't you just do:

```
 this.mapper(props);
 this.props = props;
this.getNewtonLetterText(this.props.npID[0].getContent());
```

`this.navigationSubscription` is assigned in `BeolResource`.

File: src/app/pipes/sanitize-html.pipe.ts (hunk 6 above)
Comment (tobiasschweizer) [reply src/app/pipes/sanitize-html.pipe.ts:7]: No, MathJax just assigns the content to `innerHTML`. So if we do not need it, could it be removed?

File: src/app/pipes/sanitize-url.pipe.ts (hunk 7 above)
Comment (tobiasschweizer) [reply src/app/pipes/sanitize-url.pipe.ts:7]: if it is not used anywhere, please remove it

File: src/app/resource/newton-letter/newton-letter.component.html (hunk 8 above)
Comment (tobiasschweizer) [reply src/app/resource/newton-letter/newton-letter.component.html:3]: but you use it in `getNewtonLetterText ` as well:

```typescript
private getNewtonLetterText(filename) {
	
	        this.isLoading = true;
	        // use a proxy url as described here:
	        // https://stackoverflow.com/questions/43871637/no-access-control-allow-origin-header-is-present-on-the-requested-resource-whe
	        const proxyurl = 'https://cors-anywhere.herokuapp.com/';
	        const basePath = 'http://www.newtonproject.ox.ac.uk/view/texts/normalized/';
	
	        const url = basePath + filename; // site that doesn’t send Access-Control-*
	
	        fetch(proxyurl + url) // https://cors-anywhere.herokuapp.com/https://example.com
	            .then(response => response.text())
	            .then(contents => {
	                this.getNewtonLetterBody(contents);
	                this.isLoading = false;
	            })
	            .catch(() => console.log('Can’t access ' + url + ' response. Blocked by browser?'));
	    }
```

File: src/app/pipes/sanitize-html.pipe.ts (hunk 6 above)
Comment (SepidehAlassi) [reply src/app/pipes/sanitize-html.pipe.ts:7]: I don't want to remove it because as I say with a high probability it will be needed. I will remove it later if it was not necessary for fulltext search process.

File: src/app/resource/newton-letter/newton-letter.component.html (hunk 8 above)
Comment (SepidehAlassi) [reply src/app/resource/newton-letter/newton-letter.component.html]: yeah, I set it as false always.

Comment (SepidehAlassi) [reply src/app/resource/newton-letter/newton-letter.component.html:3]: @tobiasschweizer I measured the time parser takes for the longest letter
`Call to getNewtonLetterBody took 1.2749999877996743 milliseconds.`

It is fetching the html which takes time not the parser

File: src/app/pipes/sanitize-url.pipe.ts (hunk 7 above)
Comment (SepidehAlassi) [reply src/app/pipes/sanitize-url.pipe.ts:7]: I will remove it later, if it was not necessary for the next step (fulltext search process)

File: src/app/resource/newton-letter/newton-letter.component.html (hunk 8 above)
Comment (SepidehAlassi) [reply src/app/resource/newton-letter/newton-letter.component.html:3]: @tobiasschweizer recent commit should show the progress indicator for loading the text.

File: src/app/resource/newton-letter/newton-letter.component.ts (hunk 4 above)
Comment (SepidehAlassi) [reply src/app/resource/newton-letter/newton-letter.component.ts]: @tobiasschweizer Andre added the  `this.navigationSubscription` to this code in commit 
`[96c4462](/dhlab-basel/beol/pull/108/commits/96c4462767daf383ab408aace803ab074007bffc) `
[ ](/SepidehAlassi)

so I thought navigationSubscription should be used, had not tried removing it before.

Comment (SepidehAlassi) [reply src/app/resource/newton-letter/newton-letter.component.ts]: I removed this in the last commit!

File: src/app/pipes/sanitize-html.pipe.ts (hunk 6 above)
Comment (tobiasschweizer) [reply src/app/pipes/sanitize-html.pipe.ts:7]: Ok, got it.

File: src/app/pipes/sanitize-url.pipe.ts (hunk 7 above)
Comment (tobiasschweizer) [reply src/app/pipes/sanitize-url.pipe.ts:7]: Ok, got it.

File: src/app/resource/newton-letter/newton-letter.component.html (hunk 8 above)
Comment (tobiasschweizer) [reply src/app/resource/newton-letter/newton-letter.component.html:3]: Yes, it works now!

File: src/app/resource/letter/letter.component.spec.ts [hunk 9]
```diff
@@ -37,7 +37,8 @@ describe('LetterComponent', () => {
                 ReadTextValueAsHtmlComponent,
                 ReadListValueComponent,
                 ReadTextValueComponent,
-                HanCatalogueDirective
+                HanCatalogueDirective,
+                NewtonProjectDirective
```
Comment (tobiasschweizer) [inline src/app/resource/letter/letter.component.spec.ts]: I think this is not necessary

File: src/app/resource/letter/letter.component.spec.ts [hunk 10]
```diff
@@ -14,7 +14,7 @@ import { MathJaxDirective } from '../../directives/mathjax.directive';
 import { KuiViewerModule } from '@knora/viewer';
 import { ReadTextValueComponent } from '../../properties/read-text-value/read-text-value.component';
 import { HanCatalogueDirective } from '../../directives/han-catalogue.directive';
-
+import { NewtonProjectDirective } from '../../directives/newton-project.directive';
```
Comment (tobiasschweizer) [inline src/app/resource/letter/letter.component.spec.ts]: I think this is not necessary

File: src/app/resource/newton-letter/newton-letter.component.ts (hunk 4 above)
Comment (tobiasschweizer) [reply src/app/resource/newton-letter/newton-letter.component.ts]: Ok, thanks!

**Review by tobiasschweizer — APPROVED** [review]
I just have just seen some minor things, please see my comments . Otherwise this looks good now.

**Comment by tobiasschweizer** [comment]
@SepidehAlassi So this is a step towards an interoperable edition platform. Congratulations! :-)

**Comment by SepidehAlassi** [comment]
@tobiasschweizer thanks 😄
